common/xfer/websocket.go

```diff
@@ -2,29 +2,104 @@ package xfer
 
 import (
 	"io"
+	"sync"
+	"time"
 
 	"github.com/gorilla/websocket"
 	"github.com/ugorji/go/codec"
+
+	"github.com/weaveworks/scope/common/mtime"
 )
 
-// IsExpectedWSCloseError returns boolean indicating whether the error is a
-// clean disconnection.
-func IsExpectedWSCloseError(err error) bool {
-	return websocket.IsCloseError(
-		err,
-		websocket.CloseNormalClosure,
-		websocket.CloseGoingAway,
-		websocket.CloseNoStatusReceived,
-	)
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer. Needs to be less
+	// than the idle timeout on whatever frontend server is proxying the
+	// websocket connections (e.g. nginx).
+	pongWait = 60 * time.Second
+
+	// Send pings to peer with this period. Must be less than pongWait. The peer
+	// must respond with a pong in < pongWait. But it may take writeWait for the
+	// pong to be sent. Therefore we want to allow time for that, and a bit of
+	// delay/round-trip in case the peer is busy. 1/3 of pongWait seems like a
+	// reasonable amount of time to respond to a ping.
+	pingPeriod = ((pongWait - writeWait) * 2 / 3)
+)
+
+// Websocket exposes the bits of *websocket.Conn we actually use.
+type Websocket interface {
+	ReadMessage() (messageType int, p []byte, err error)
+	WriteMessage(messageType int, data []byte) error
+	ReadJSON(v interface{}) error
+	WriteJSON(v interface{}) error
+	Close() error
+}
+
+type pingingWebsocket struct {
+	pinger    *time.Timer
+	readLock  sync.Mutex
+	writeLock sync.Mutex
+	conn      *websocket.Conn
+}
+
+// Ping adds a periodic ping to a websocket connection.
+func Ping(c *websocket.Conn) Websocket {
+	p := &pingingWebsocket{conn: c}
+	p.conn.SetPongHandler(p.pong)
+	p.pinger = time.AfterFunc(pingPeriod, p.ping)
+	p.ping()
+	return p
+}
+
+func (p *pingingWebsocket) ping() {
+	p.writeLock.Lock()
+	defer p.writeLock.Unlock()
+	if err := p.conn.WriteControl(websocket.PingMessage, nil, mtime.Now().Add(writeWait)); err != nil {
+		p.Close()
+	}
+	p.pinger.Reset(pingPeriod)
+}
+
+func (p *pingingWebsocket) pong(string) error {
+	p.conn.SetReadDeadline(mtime.Now().Add(pongWait))
+	return nil
+}
+
+// ReadMessage is a helper method for getting a reader using NextReader and
+// reading from that reader to a buffer.
+func (p *pingingWebsocket) ReadMessage() (int, []byte, error) {
+	p.readLock.Lock()
+	defer p.readLock.Unlock()
+	messageType, b, err := p.conn.ReadMessage()
+	if err == nil {
+		err = p.conn.SetReadDeadline(mtime.Now().Add(pongWait))
```

Comment:
It's the next time we need to see something from the other side before disconnecting.